plugins/optimization-detective/class-od-tag-visitor-registry.php

```diff
@@ -37,9 +37,20 @@ final class OD_Tag_Visitor_Registry implements Countable, IteratorAggregate {
 	 *
 	 * @param string   $id                   Identifier for the tag visitor.
 	 * @param callable $tag_visitor_callback Tag visitor callback.
+	 * @param array<mixed> $dependencies array with ID as key and Callable as value
 	 */
-	public function register( string $id, callable $tag_visitor_callback ): void {
-		$this->visitors[ $id ] = $tag_visitor_callback;
+	public function register( string $id, callable $tag_visitor_callback, array $dependencies = array() ): void
+	{
+		if (!empty($dependencies)) {
+			foreach ($dependencies as $dependency_id => $dependency_callback) {
+				if (!self::is_registered($dependency_id)) {
+					$this->visitors[$dependency_id] = $dependency_callback;
+				}
+			}
+		}
+		if (!self::is_registered($id)) {
+			$this->visitors[$id] = $tag_visitor_callback;
+		}
 	}
```

Comment:
I was thinking rather that this would work more like how `WP_Script` works, where you can register a script (or tag visitor) with a given handle (or ID) and supply dependencies which are a list of strings (i.e. handles/IDs). Then later when obtaining the list of tag visitors to apply, it would recursively look up the registered tag visitors to make sure that those which have no dependencies run first, followed by those which have dependencies.

Reply:
I can see a way of changing this to just passing the className and the dependencies as a series of class names
and then getting the ID from the class 

Or just building an array values and then load them in the first call to get_registered

Part of the problem is I am not sure what the code is doing

Reply:
The tag visitor registry is a collection of callables which are invoked when iterating over the tags via HTML Tag Processor. A tag visitor may be an invokable class or it may be a closure or a function or any other `callable`. See the Enhanced Responsive Images for an example tag visitor which is a regular PHP function and not a class:

https://github.com/WordPress/performance/blob/bc37748d38c6e696f085f8ffedb19389d3198151/plugins/auto-sizes/optimization-detective.php#L13-L52

So in the same way as with `WP_Scripts` where registered scripts have handles, here too tag visitors (callables) have IDs. As with script dependencies being able to register a script which depends on another script by passing an array of dependency handles, so too as seen in Enhanced Responsive Images there should be a way to register a tag visitor so that it runs after other tag visitors have done.